docs/proposals/22-autoscaling-kube-apiserver-via-independent-hpa-and-vpa.md

```diff
@@ -0,0 +1,322 @@
+these API extension points. 
+
+### Element: HPA
+ShootKapi is scaled horizontally via HPA operating on the aforementioned average kube-apiserver request rate custom
+metric. Scale-in is delayed by a stabilization window. This avoids unnecessary flapping, the effects of which would
+be further amplified by VPA-driven evictions, and serves as a general precaution against HPA-VPA resonance,
+by ensuring that the two autoscalers operate with different frequency.
+
+The scaling threshold value for HPA is initially set to a conservative (lower-biased) estimate of the requests/replica
+ratio observed with the existing solution. It will be then gradually fine-tuned to the maximum value which does not
+impact quality of service.
+
+This proposal is the first stage of an incremental, two stage approach. It aims primarily at resolving the existing
+stability issues inherent to HVPA and enabling (but not executing) scaling efficiency optimizations. The second stage
+will be focused solely on further fine-tuning scaling efficiency to optimize ShootKapi compute cost over HVPA.
+Details follow in the [Discussion and Limitations](#discussion-and-limitations) section.
+
+### Element: VPA
+A standard VPA operates on ShootKapi. Its primary purpose is to improve resource efficiency over what's possible via
+HPA's coarse, replica-granularity scaling, by shrinking replicas down to match actual utilization.
+
+This improves over the existing solution by having vertical scaling active the entire time and not only during certain
+periods.
+
+### Transition Strategy
+The following steps will be executed over time, and provide a non-disruptive transition from the existing implementation
+to the one hereby proposed.
+
+1. Proposed scaling approach will be deployed behind a feature gate and disabled by default. Existing HVPA feature flag
+   will be preserved at this point, as its use is not limited to kube-apiserver scaling.
+2. New feature gate gradually enabled until all ShootKapi workloads scaled by new approach (Note: this step applies to
+   individual garden/seed instances, and not to the code base as a whole)
+3. At this point, the preexisting HVPA feature flag has no effects on ShootKapi scaling (Note: this step applies to
+   individual garden/seed instances, and not to the code base as a whole)
+4. New feature flag promoted and eventually removed
+5. (Outside the scope of this proposal) use of HVPA removed for other workloads and the preexisting HVPA flag removed
+
+## Discussion and Limitations
+#### Prometheus-metrics-adapter precludes the use of sources of custom metrics
+With current K8s design, only one API extension can serve custom metrics. Per this GEP, the custom metrics extension
+point is occupied by prometheus-metrics-adapter.
+
+#### Heavyweight Prometheus infrastructure feeds signal back into cluster control loop
+The proposed signal flow brings two Prometheus instances (shoot prometheus and seed aggregate prometheus) into the shoot
```

Comment:
Done and will be available with next push, but I'd like to explain, what I was driving at by using "seed aggregate prometheus":

Gardener's monitoring setup is complex enough so for those who are new to the project, or maybe even those who just do not follow developments closely, it is easy to lose track of its structure. I think for those people it's easier when prometheus instances are prefixed with indication whether they run as part of seed or shoot, e.g:
- seed aggregate prometheus
- seed cache prometheus
- seed main prometheus
- shoot control plane prometheus (And thanks for adding the "control plane" in your comment. I don't know if that was your intent, but it's exactly in the spirit of my concern. For someone who is in doubt whether "shoot prometheus" may be referring to one running inside the shoot, that's a welcome clarification)

I know that those aren't the correct names, and you are pointing me the correct ones. So I am not advocating that I use the incorrect ones in this PR. I just wanted to provide a monitoring-outsider perspective, because it may not be obvious to those who know the monitoring setup well.